### ISSUE TITLE
histogram with Freedman Diaconis estimator crashes on small numbers

### ISSUE THREAD

**Issue by asanakoy** [numpy/numpy#8203]
`np.histogram` crashes with `MemoryError` when the input contains very small numbers.

```
In [13]:  print(sys.version)
2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2]

In [14]: print(np.__version__)
1.11.2

In [15]: x = np.array([ 2, 2, 2 - 1e-15, 2 - 1e-15, 1], dtype=np.float64)

In [16]: np.histogram(x, bins='fd')
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-16-4ee7362b8a29> in <module>()
----> 1 np.histogram(x, bins='fd')

/home/artem/.local/lib/python2.7/site-packages/numpy/lib/function_base.pyc in histogram(a, bins, range, normed, weights, density)
    562 
    563         # Initialize empty histogram
--> 564         n = np.zeros(bins, ntype)
    565         # Pre-compute histogram scaling factor
    566         norm = bins / (mx - mn)

MemoryError: 

```

I did some debugging and figured out that in `numpy.lib.function_base._hist_bin_fd`  returns bin width 1.29852472695e-15

```
def _hist_bin_fd(x):  # x: array([ 2, 2, 2 - 1e-15, 2 - 1e-15, 1])
    iqr = np.subtract(*np.percentile(x, [75, 25]))  # iqr: 1.11022302463e-15
    return 2.0 * iqr * x.size ** (-1.0 / 3.0)  # returns 1.29852472695e-15
```

I think that in `/home/artem/.local/lib/python2.7/site-packages/numpy/lib/function_base.py:533` it would be good to check that we get a reasonable number of bins (<= then number of points in the array).

**Comment by asanakoy** [comment]
This kind of small numbers could appear due to precision errors in float64 representation.
I understand that in some applications people work with huge amounts of memory and can have small bin width on purpose. But there is no sense in having 10^8 more bins than number of samples in the array.

At list such a warning can be useful.

```
def _hist_bin_fd(x):
    iqr = np.subtract(*np.percentile(x, [75, 25]))
    h = 2.0 * iqr * x.size ** (-1.0 / 3.0)
    if (x.max() - x.min()) / h > 1e8 * x.size:
        warnings.warn("Bin width estimated with the Freedman-Diaconis rule is very small"
                      " (= {})".format(h), RuntimeWarning, stacklevel=2)
    return h
```

**Comment by shoyer** [comment]
Maybe we reasonably limit the number of bins to less than the array size?

CC  @nayyarv @madphysicist

**Comment by madphysicist** [comment]
Not in the general case, no. Imagine a dataset with two clumps of data separated by a huge gap for an example of why not. It may, however, be worth considering merging together empty bins. To remove what I imagine to be the most common cause for the issue you are describing.

**Comment by asanakoy** [comment]
@shoyer This is not always good idea. Sometimes you would like to have more bins than teh array size. 
For example when you have a sum of 2 gaussians. first with mean=0, std=1, and the second with mean=100, std=1. And the array size = 100 (50 samples drown from the first gaussian, 50 from the second). Then with 100 equally sized bins you won't be able to approximate it, as the bin size will be 1.

**Comment by madphysicist** [comment]
Another thing of note is that the IQR is exactly zero if you leave out the 1e-15 and just do `x = np.array([ 2, 2, 2, 2, 1], dtype=np.float64)`. My gut response would be to recommend using a different estimator for this sort of data, but I will think for a bit and try to come up with a more reasonable solution.

**Comment by asanakoy** [comment]
That's true. Because IQR is the difference between 25 and 75 percentiles. 
The funny thing is that I faced this Error when was working with some data, that didn't imply such a big precision. Occasionally I got a 30-element array with a lot of 1-s , 1.7-s and 2-s. And some of them differed from the integer only on a small delta, probably because of floating point calculations and rounding errors

**Comment by nayyarv** [comment]
The most reasonable suggestion I can think of is doing a check that the standard deviation and the IQR are within a similar order of magnitude and reverting to the standard deviation in extreme cases. 

Something along the lines of 

```python
iqr = iqr = np.subtract(*np.percentile(x, [75, 25]))
stddev = np.std(x)

if np.abs(np.log10(iqr) - np.log10(stddev)) > 2:
   iqr = stddev
```

The standard deviation is more susceptible to outliers but has less variance since it depends on all the data points. It should always be very close in magnitude to the IQR.

**Comment by nayyarv** [comment]
Or alternatively round to 3sf when passing to the estimators.

**Comment by madphysicist** [comment]
I was thinking that we could fail over to std instead of iqr if the bin count was too high rather than pre-computing the std for the comparison. Off the top of my head, we already have the min and max available.

**Comment by madphysicist** [comment]
And raise a warning about the data having too fine a structure or something.

**Comment by asanakoy** [comment]
But the question what is the high number of bins? It depends on size of the available RAM. For my linux machine with 8GB RAM, number if bins 10**8 made me go to swap.

**Comment by madphysicist** [comment]
Good point. Number of bins vs number of data points might be a better metric. 10**8 bins for a histogram of 10**10 values is probably fine, but not for 10 values.

**Comment by asanakoy** [comment]
The simple solution that fixed this bug for me is just force conversion from float64 to float32 when using FD estimator.

with `n_bins / n_val` is kind of tricky. I'm not sure, but maybe one will want to have 10**7 bins and 10**4 points and the ratio in this case will be 10**3. 

I would suggest something like: 

    if  n_bins > 1e8 and n_bins / n_val > 10e2:
        x = x.astype(np.float32)

**Comment by madphysicist** [comment]
I was thinking something even simpler:
```
if n_bins / n_val > 1e6:
    warnings.warn(...)
    # recompute with std instead of iqr.
```

**Comment by madphysicist** [comment]
With float32 your epsilon would be around 1e-5 instead of 1e-15, but you could still trigger the undesired behavior.

**Comment by asanakoy** [comment]
As I noticed usually I don't get this hard rounding error with float. At least for the test that I wrote in the first post it worked.
But maybe you are right.

**Comment by anntzer** [comment]
If merging together empty bins is a possibility, then the number of bins could at least be limited to 2x the number of data points.
